tests/returntypes/test_spsa_gradient_new.py

```diff
@@ -0,0 +1,1052 @@
+        )
+        assert isinstance(res[1][0], numpy.ndarray)
+        assert np.allclose(
+            res[1][1],
+            [
+                -(np.cos(x / 2) ** 2 * np.sin(y)) / 2,
+                (np.cos(x / 2) ** 2 * np.sin(y)) / 2,
+                (np.sin(x / 2) ** 2 * np.sin(y)) / 2,
+                -(np.sin(x / 2) ** 2 * np.sin(y)) / 2,
+            ],
+            atol=tol,
+            rtol=0,
+        )
+        assert isinstance(res[1][1], numpy.ndarray)
+
+
+@pytest.mark.parametrize(
+    "sampler, num_directions, atol", [(_rademacher_sampler, 4, 0.5), (coordinate_sampler, 2, 1e-3)]
+)
+class TestSpsaGradientDifferentiation:
+    """Test that the transform is differentiable"""
+
+    @pytest.mark.autograd
+    def test_autograd(self, sampler, num_directions, atol):
+        """Tests that the output of the SPSA gradient transform
+        can be differentiated using autograd, yielding second derivatives."""
+        dev = qml.device("default.qubit.autograd", wires=2)
+        params = np.array([0.543, -0.654], requires_grad=True)
+        np.random.seed(42)
+
+        def cost_fn(x):
+            with qml.tape.QuantumTape() as tape:
+                qml.RX(x[0], wires=[0])
+                qml.RY(x[1], wires=[1])
+                qml.CNOT(wires=[0, 1])
+                qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+            tape.trainable_params = {0, 1}
+            tapes, fn = spsa_grad(tape, n=1, num_directions=num_directions, sampler=sampler)
+            jac = np.array(fn(dev.batch_execute(tapes)))
+            if sampler == coordinate_sampler:
+                jac *= 2
+            return jac
+
+        res = qml.jacobian(cost_fn)(params)
+        x, y = params
+        expected = np.array(
+            [
+                [-np.cos(x) * np.sin(y), -np.cos(y) * np.sin(x)],
+                [-np.cos(y) * np.sin(x), -np.cos(x) * np.sin(y)],
+            ]
+        )
+
+        assert np.allclose(res, expected, atol=atol, rtol=0)
+
+    @pytest.mark.autograd
+    def test_autograd_ragged(self, sampler, num_directions, atol):
+        """Tests that the output of the SPSA gradient transform
+        of a ragged tape can be differentiated using autograd, yielding second derivatives."""
+        dev = qml.device("default.qubit.autograd", wires=2)
+        params = np.array([0.543, -0.654], requires_grad=True)
+        np.random.seed(42)
+
+        def cost_fn(x):
+            with qml.tape.QuantumTape() as tape:
+                qml.RX(x[0], wires=[0])
+                qml.RY(x[1], wires=[1])
+                qml.CNOT(wires=[0, 1])
+                qml.expval(qml.PauliZ(0))
+                qml.probs(wires=[1])
+
+            tape.trainable_params = {0, 1}
+            tapes, fn = spsa_grad(tape, n=1, num_directions=num_directions, sampler=sampler)
+            jac = fn(dev.batch_execute(tapes))
+            if sampler == coordinate_sampler:
+                jac = tuple(tuple(2 * _j for _j in _jac) for _jac in jac)
```

Comment:
What happens with the coordinate sampler for it to miss a factor 2?.

Reply:
The concept of `SPSA` is to sample many directions, perform 1D finite differences for those, and average them to get an estimate for the gradient. If we want to make `SPSA` to actually compute the exact same gradient as `finite_diff`, i.e. sample each parameter direction successively, we instead want to compute the finite diff with the first direction being `[1, 0....0]` , the second `[0, 1, 0...0]` and so on, and sum them up.
The `mean` sums them up, but also divides by the number of directions, which we need to revert here, in order to just get the sum.

Reply:
Good to know, is it possible to add a comment about it somewhere?

Reply:
Yes, I'll add this technical detail to the docstring of `spsa_grad` itself!